### ISSUE TITLE
`git` resource type to support version as [sha1, path]

### ISSUE THREAD

**Issue by evanchaoli** [concourse/concourse#3189]
## What challenge are you facing?

Use Case: We have a repo. In this repo, each subfolder contains code of a docker-image. We have a pipeline that auto build docker-image upon code change.

`git` resource has supported to filter change by dir. We have using this feature. In the pipeline, we define a `git` resource for each sub-folder. This way works, but the problem is, whenever we add a new sub-folder, we have to modify the pipeline code to add a `git` resource as well as a new job, which is totally duplicate effort.

As `git` resource has already had `paths` and `ignore_paths` parameters, we can add a new boolean parameter, say `version_with_path`, when this parameter is true, then git resource will return version as [sha1, path_name]. If multiple paths have changes, `check` should return multiple versions:

    [
       {"version": [sha1, path1]},
       {"version": [sha1, path2]},
       ...
    ]

When `version_with_path` is true, we can introduce a new parameter, say `version_path`. If `path` is not specified, then `paths` defaults to all sub-folders of `version_path`. And `version_path` defaults to root of the repo.

In summary, when `version_with_path` is true, the behavior is:

1. Filter out changes of `ignore_paths`.
2. If `paths` is not given, populate `paths` with all subfolders of `version_path`.
3. If some `paths` contain changes, then return a serial of versions as [sha1, path]
4. `in` will generate one more file, say `path`, that tells which path this version is for.

Thoughts?

## Are you interested in implementing this yourself?

Yes we can.

**Comment by vito** [comment]
🤔 This sounds like it should be its own resource type. And when #1707 is done it could even use spaces for each path, which would be pretty cool. But I don't think this should be added to the `git` resource right now as it's already turning into a bit of a swiss-army knife.
